File: packages/guards/src/stereotype-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface StereotypeDetectOptions { action: 'block' | 'warn'; }

const STEREOTYPE_PATTERNS: RegExp[] = [
  /\b(?:all|every|typical)\s+(?:women|men|boys|girls|asians?|blacks?|whites?|latinos?|muslims?|christians?|jews?)\s+(?:are|always|never|can'?t)\b/gi,
  /\b(?:women|men)\s+(?:belong|should\s+(?:stay|be))\s+(?:in|at)\s+(?:the\s+)?(?:home|kitchen|office|work)\b/gi,
  /\b(?:too\s+(?:old|young)\s+to|(?:old|young)\s+people\s+(?:can'?t|don'?t|shouldn'?t))\b/gi,
  /\b(?:disabled|handicapped)\s+people\s+(?:can'?t|shouldn'?t|don'?t)\b/gi,
  /\b(?:real\s+(?:men|women))\s+(?:don'?t|always|should)\b/gi,
  /\b(?:that'?s\s+(?:so\s+)?(?:gay|girly|manly|ghetto|ratchet))\b/gi,
  /\b(?:you\s+(?:people|guys|lot))\b/gi,
];

export function stereotypeDetect(options: StereotypeDetectOptions): Guard {
  return { name: 'stereotype-detect', version: '0.1.0', description: 'Detect stereotyping language about groups', category: 'content', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const matched: string[] = [];
      for (const p of STEREOTYPE_PATTERNS) { const re = new RegExp(p.source, p.flags); const m = re.exec(text); if (m) matched.push(m[0]); }
      const triggered = matched.length > 0;
      return { guardName: 'stereotype-detect', passed: !triggered, action: triggered ? options.action : 'allow',
        message: triggered ? `Stereotype detected: "${matched[0]}"` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'Text contains stereotyping language about demographic groups' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/stereotype_detect.py
"""Detect stereotype patterns."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_GROUP_WORDS = (
    r"(?:women|men|boys|girls|whites|blacks|"
    r"asians|latinos|hispanics|muslims|jews|"
    r"christians|hindus|immigrants|elderly|"
    r"millennials|boomers|gen\s*z|teenagers|"
    r"disabled|lgbtq|gay|lesbian|trans)"
)

_STEREOTYPE_PATTERNS = [
    re.compile(
        r"typical\s+" + _GROUP_WORDS,
        re.IGNORECASE,
    ),
    re.compile(
        _GROUP_WORDS + r"\s+always",
        re.IGNORECASE,
    ),
    re.compile(
        _GROUP_WORDS + r"\s+never",
        re.IGNORECASE,
    ),
    re.compile(
        _GROUP_WORDS + r"\s+can'?t",
        re.IGNORECASE,
    ),
    re.compile(
        r"all\s+" + _GROUP_WORDS + r"\s+are",
        re.IGNORECASE,
    ),
    re.compile(
        _GROUP_WORDS
        + r"\s+(?:should|belong|need\s+to)",
        re.IGNORECASE,
    ),
]


class _StereotypeDetect:
    def __init__(
        self, *, action: str = "warn"
    ) -> None:
        self.name = "stereotype-detect"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        found: list[str] = []

        for pat in _STEREOTYPE_PATTERNS:
            match = pat.search(text)
            if match:
                found.append(match.group())

        elapsed = (
            time.perf_counter() - start
        ) * 1000

        if not found:
            return GuardResult(
                guard_name="stereotype-detect",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        return GuardResult(
            guard_name="stereotype-detect",
            passed=False,
            action=self.action,
            message="Stereotype pattern detected",
            latency_ms=round(elapsed, 2),
            details={
                "matched": found,
                "reason": (
                    "Text contains generalizations"
                    " about a group"
                ),
            },
        )


def stereotype_detect(
    *, action: str = "warn"
) -> _StereotypeDetect:
    return _StereotypeDetect(action=action)
